### PR TITLE
claude-code-chat-browser: YAML escape fix — handle colons, hashes, tabs, bool literals in frontmatter export

### LINKED ISSUE #102: claude-code-chat-browser: YAML escape fix — handle colons, hashes, tabs, bool literals in frontmatter export
## Calendar Day

Wednesday, July 2, 2026 (**PR 2 of 2**)

## Planned Effort

**3 story points** (Medium–High) — sprint item **#6**

**Companion PR:** Wednesday PR 1 (SessionMetadataDict #9) — independent; can land in either order.

## Problem

The `_escape_yaml()` helper in `utils/md_exporter.py` does not handle several YAML-special characters and patterns:

- **Colons followed by space** (`: `) — interpreted as key-value separator, corrupts frontmatter
- **Hash/pound signs** (`#`) — interpreted as inline comments, truncates values
- **Tab characters** — YAML spec forbids tabs for indentation
- **YAML boolean literals** (`true`, `false`, `yes`, `no`, `on`, `off`) — unquoted values parsed as booleans

Session titles and project paths from Claude Code regularly contain colons (e.g. `C:\Users\...`, `Fix: handle edge case`) and hashes. The resulting YAML frontmatter is malformed and fails to parse in downstream consumers.

## Goal

One merged PR that makes exported YAML frontmatter round-trip cleanly through `yaml.safe_load()` for all real-world session metadata inputs, using always-double-quote scalars as the primary fix.

### Touch points

- `utils/md_exporter.py` — `_escape_yaml()`: always double-quote YAML scalar values in frontmatter; escape internal `"` as `\"`; use block scalar (`|` or `>`) for multiline values
- `tests/` — `test_yaml_frontmatter_roundtrip`: extract YAML between `---` delimiters, `yaml.safe_load()`, assert dict matches original metadata
- `tests/` — Hypothesis `@given(st.text())` property test for round-trip fidelity

### Out of scope

- Changes to Markdown body export format
- JS YAML parser compatibility tests (Python round-trip is sufficient)

## Acceptance Criteria

- [ ] `_escape_yaml()` produces valid YAML for values containing `: `, `#`, `\t`, and boolean literals
- [ ] Double quotes within values escaped as `\"` before wrapping
- [ ] Multiline values use YAML block scalar syntax
- [ ] Exported frontmatter round-trips through `yaml.safe_load()`
- [ ] Hypothesis property-based test asserts round-trip fidelity
- [ ] Existing export tests pass with properly escaped output
- [ ] Full `pytest`, `mypy --strict`, and `ruff` pass
- [ ] PR approved by at least 1 reviewer

## Verification

```powershell
cd C:\Users\Jasen\CppAliance\claude-code-chat-browser
.\.venv\Scripts\Activate.ps1
pytest tests/ -k yaml -q
pytest -q
```

Manual: export session with title `Fix: handle edge case` — confirm frontmatter parses via `yaml.safe_load()`.

### PR DESCRIPTION
**Pull request by clean6378-max-it**
Closes #102 

## Summary
* Refactor YAML frontmatter export to use `_session_frontmatter_dict()` + `_append_yaml_value()`
* `_escape_yaml()` always emits double-quoted scalars with proper escaping for `"`, `\`, tabs, newlines, control chars, and non-printable Unicode
* `models_used` and `service_tiers` emit as YAML sequences (not comma-joined strings)
* Nested mapping keys (e.g. tool names with colons) are quoted when not plain identifiers
* Add `test_yaml_frontmatter_roundtrip` and Hypothesis `@given(st.text())` property tests
* Add `PyYAML>=6.0` to `requirements-dev.txt`
Sprint item **#6** (3 pt). Independent of PR #106 (SessionMetadataDict).

## Problem

Unquoted frontmatter values broke on colons, `#` comments, tabs, and YAML boolean literals (`true`, `off`, etc.) in real session titles and paths.

## Test plan

- [x] `pytest tests/test_md_exporter_yaml.py` — 5 passed
- [x] `pytest -q` — 424 passed, 13 skipped
- [x] `mypy api utils models` — pass
- [x] `ruff check .` — pass
- [ ] Manual: export session titled `Fix: handle edge case` — confirm `yaml.safe_load()` parses frontmatter

## Summary by CodeRabbit

* **New Features**
  * Markdown exports now generate richer, consistently ordered YAML frontmatter metadata.
* **Bug Fixes**
  * Improved YAML escaping/serialization for exported metadata, including multiline titles and special characters (tabs, `#`, quotes, backslashes, newlines).
* **Tests**
  * Added automated round-trip and fuzz coverage to verify YAML frontmatter correctness and escaping behavior.
* **Chores**
  * Updated development dependencies to include YAML parsing support for tests.

### CODE REVIEW

**Comment by timon0305** [comment]
The PR body claims "Multiline printable strings use |- block scalars; non-printable multiline falls back to quotes escapes." But _escape_yaml always emits a double-quoted scalar with \n for newlines - there's no block-scalar code anywhere test_multiline_title_uses_quoted_scalar even asserts the quoted from. 
The quoted approach is fine and round-trips, so just fix the description - right now it describes behavior the code doesn't have

File: utils/md_exporter.py [hunk 1]
```diff
@@ -25,63 +26,70 @@ def session_to_markdown(session: SessionDict, stats: SessionStatsDict | None = N
...
+    if meta.get("version"):
+        data["claude_code_version"] = meta["version"]
+    if meta.get("permission_mode"):
+        data["permission_mode"] = meta["permission_mode"]
     if meta.get("service_tiers"):
-        lines.append(f"service_tiers: {', '.join(meta['service_tiers'])}")
-    lines.append(f"message_count: {len(session['messages'])}")
-    if meta["compactions"] > 0:
-        lines.append(f"compactions: {meta['compactions']}")
+        data["service_tiers"] = ", ".join(meta["service_tiers"])
```
Comment (timon0305) [inline utils/md_exporter.py]: These round-trip as a single string("a", "b"), so a YAML consumer can't treat them as a list. This is a preserved from the old code, and the round-trip test passes because both sides use the joined string - but since this PR is specifically about producing correct YAML frontmatter, emitting them as real YAML sequences would be the more useful representation

**Review by timon0305 — APPROVED** [review]

**Comment by clean6378-max-it** [comment]
> The PR body claims "Multiline printable strings use |- block scalars; non-printable multiline falls back to quotes escapes." But _escape_yaml always emits a double-quoted scalar with \n for newlines - there's no block-scalar code anywhere test_multiline_title_uses_quoted_scalar even asserts the quoted from. The quoted approach is fine and round-trips, so just fix the description - right now it describes behavior the code doesn't have

Good catch — block scalars were removed in df37c92,  updated the PR description

File: utils/md_exporter.py (hunk 1 above)
Comment (clean6378-max-it) [reply utils/md_exporter.py]: Done. models_used and service_tiers are kept as lists in the canonical frontmatter dict and serialized as YAML sequences.

Round-trip tests confirm yaml.safe_load() returns real lists.

**Review by wpak-ai — APPROVED** [review]
